### DISCUSSION TITLE
Maintain Diagnostic on Non File Node

### DISCUSSION THREAD

**Discussion by kapishsingla** [eclipse-langium/langium#1741] [Q&A]
Hi,

In theia, I have extended the `FileNavigatorWidget `to serve the purpose of showing the components of document as a non-file Node in Navigator. 

<img width="134" alt="image" src="https://github.com/user-attachments/assets/53a6ccde-b30b-481d-9807-fd09b39be64d">

Theia currently uses `LSP` with Langium documents to generate diagnostics and display markers on files in the Navigator, working at the document level. I want to adjust this so diagnostics are maintained and retrieved based on non-file nodes. This way, I can get diagnostics specific to a non-file node and display markers only on that particular node.

Could you please suggest any approach for the same?

**Comment by spoenemann** [comment]
Hi @kapishsingla! I would create a custom JSON-RPC message where you send exactly the information you need in your Navigator tree. Here's an example for this:

https://github.com/eclipse-langium/langium/blob/21b105164186dfff6a4763c21d450931e2200e5b/examples/statemachine/src/language-server/main-browser.ts#L25-L38

The example sends the whole serialized AST. In your case, you rather need information about validation results and to which nodes they belong.

Then you need corresponding code on the VS Code extension side to register a message handler on the JSON-RPC connection.
